### ISSUE TITLE
Support CUDA and backpropagation in `torch.orgqr`

### ISSUE THREAD

**Issue by toshas** [pytorch/pytorch#50104]
## 🚀 Feature

As of PyTorch 1.7, `torch.orgqr` only supports CPU tensors and does not implement gradients. 
Feature proposal: add support for CUDA tensors and gradients wrt inputs.

## Motivation

This function is required to perform orthogonal parameterization of matrices (not to be confused with orthogonal initialization). Given an input matrix of shape `d x r` with `d >= r`, `orgqr` produces an orthogonal matrix of the same shape `d x r`. The output spans the manifold of Stiefel frames of size `d x r`, and hence, provides precise orthogonal constraints by construction. This can be used further, e.g. as a weight matrix in an RNN transition layer or any other place requiring strict orthogonality. More motivation can be found here and references therein: https://github.com/pytorch/pytorch/issues/42243
Without GPU, the function is slow. Without gradients, parameterization becomes impossible. Once the requested feature is implemented, and once this PR https://github.com/pytorch/pytorch/pull/33344 is merged, it will become possible to perform precise orthogonal parameterization of weight matrices in Linear, all flavors of Convolutional, and Embedding layers.

## Pitch

`torch.orgqr` to support the two requested properties.
As hinted by @Lezcano, MAGMA already implements a GPU side of the function: https://www.icl.utk.edu/~mgates3/docs/magma-proto.html
The gradients can be taken from my package https://github.com/toshas/torch-householder/blob/master/torch_householder/householder.cpp#L58

## Alternatives

The code from my package https://github.com/toshas/torch-householder implements an efficient drop-in replacement for `torch.orgqr` satisfying the requested properties, and I don't mind donating this code. The forward pass might be slower than that of MAGMA -- no comparison was done.

## Additional context

None so far

cc @ngimel @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk

**Comment by mruberry** [comment]
Thank you for this suggestion, @toshas. Can you comment on why you're interested in torch.orgqr, in particular, vs. the new torch.linalg.qr? See its documentation here: https://pytorch.org/docs/master/linalg.html#torch.linalg.qr.

**Comment by toshas** [comment]
Great question @mruberry! QR decomposition takes an arbitrary matrix `W` as an input, and returns an orthogonal matrix `Q`, and a triangular matrix `R`. This function is useful on its own, but it does not help to solve parameterization of the manifold, which is the "inverse" problem in some sense. 

Say we want a matrix `W` (of shape `d x r`)(which could be a weight matrix of a layer) be orthogonal - how do we do that? One way is to parameterize it as `W = torch.nn.Parameter(torch.randn(d, r))` and add a loss term, penalizing for the deviation of the matrix from orthogonality constraints: `loss = loss + lambda * (W.T.mm(W) - torch.eye(r)).abs().mean()` . This is how it is done most of the times (orthogonal regularization). However, this way is problematic:
- Nothing guarantees that `W` will be maintained orthogonal throughout training process, it can be anything, depending on the interactions of the parameterized model, optimizer, and the dataset;
- The extra hyperparameter `lambda` adds unnecessary complexity/degrees of freedom.

The proposed way to maintain precise constraints of the desired manifold (orthogonality of `W` in our case) is to perform direct (generative) parameterization of the manifold. For the manifold of Stiefel frames, a number of approaches exist: Givens rotations, Matrix exponential, and Householder transformation, to name a few. 

A number of works ([example](https://pubs.acs.org/doi/pdf/10.1021/acs.jpca.5b02015)) conclude that Householder transformation is the cheapest in terms of compute and memory. In my package I made a [comparison](https://github.com/toshas/torch-householder#speed) of this transformation with Matrix exponential; while not the fastest for square matrices, it shines with non-square matrices, large batch sizes, memory requirements with all matrix sizes, and numerical precision with all matrix sizes.

To conclude, the requested feature facilitates the Householder transformation and parameterization of matrices, which will be especially easy to use when this PR https://github.com/pytorch/pytorch/pull/33344 is merged.

**Comment by Lezcano** [comment]
A more mathematical take on the problems of QR vs orgqr:

The current QR implementation does not implement backpropagation whenever the input is not full rank. This is not because the current implementation is lacking, but because the gradient is not well defined for these inputs (there are many possible gradients really). This could be solved by a careful implementation of QR with pivoting, but even then there are problems at zero and it has some issues, like, if you start on a matrix with rank `r`, the gradients will never increase the rank of this matrix. See the discussion in:
https://github.com/pytorch/pytorch/issues/42792

On the other hand, this operation allows to define a map from `R^{nxk}` onto the set of orthogonal matrices, without having the problems that the full QR decomposition has.

**Comment by mruberry** [comment]
Thank you, @toshas, @Lezcano. Also, let me cc @jbschlosser who's looked at better support for orthogonalization constraints recently. 

From your analysis I will tentatively suggest that we would accept a PR implementing this. More correctly, it would be interesting to see two (maybe three) PRs: one adding CUDA support, and another autograd support. 

fyi we're also currently working on improving PyTorch's linear algebra support. That focus has mostly been on the development of the new torch.linalg module and getting to feature parity with NumPy. After that, however, we intend to extend our review to other linear algebra functions in PyTorch, like orgqr, too.  

What do you think, @IvanYashchuk?

**Comment by IvanYashchuk** [comment]
I think it's a good idea not to constrain ourselves to support only what NumPy or SciPy supports. Especially in this case when LAPACK, MAGMA, cuSOLVER already provide a ready routine to use.
Even though LAPACK API is well-known to some people, this new function should probably have a different more descriptive name as we are not constrained by the old Fortran 6 character function name convention. `torch.orgqr` and `torch.geqrf` are the only functions exposing the LAPACK function name.

Steps implementing this would be:
1. Port `orgqr` from TH to ATen, adding also support for complex dtype inputs (using LAPACK's `ungqr`)
2. Add autograd support
3. Add CUDA support using MAGMA or cuSOLVER depending on what is more efficient

**Comment by toshas** [comment]
@IvanYashchuk @mruberry 
May I suggest the following example be added to the function documentation, given that parameterization functionality is merged now? It demonstrates the simplest usage of this function to implement exact orthogonality constraints in linear layers and directs the reader to Sec.7 of our recent publication: https://arxiv.org/abs/2103.04217. It elaborates how to make use of the batch dimension, parameterize multiple matrices of different shapes within the same batch, and more.

After the line https://github.com/pytorch/pytorch/blob/2d14bf3dccc1ffe7781405fcb2805ff5a58fbd69/torch/linalg/__init__.py#L384
```
The second example shows the orthonormal parameterization of a weight matrix.
Parameterizing multiple matrices of different sizes in a single batch and reduced parameterizations are discussed in Sec.7 of https://arxiv.org/abs/2103.04217.
```

After the line https://github.com/pytorch/pytorch/blob/2d14bf3dccc1ffe7781405fcb2805ff5a58fbd69/torch/linalg/__init__.py#L423
```
Example 2: Exact orthonormal constraints on weight matrices:

>>> class Orthonormal(torch.nn.Module):
>>>    def __init__(self, d, r):
>>>        super().__init__()
>>>        self.register_buffer('eye', torch.eye(d, r))
>>>    def forward(self, hh):
>>>        hh = hh.tril(-1) + self.eye
>>>        tau = 2 / (hh * hh).sum(dim=0)
>>>        return torch..linalg.householder_product(hh, tau)
>>>
>>> cin, cout = 128, 256
>>> layer = torch.nn.Linear(cin, cout)
>>> torch.nn.utils.parametrize.register_parametrization(layer, 'weight', Orthonormal(cout, cin))
ParametrizedLinear(
  in_features=128, out_features=256, bias=True
  (parametrizations): ModuleDict(
    (weight): ParametrizationList(
      (0): Orthonormal()
    )
  )
)
>>> torch.allclose(layer.weight.T.mm(layer.weight), torch.eye(cin), atol=1e-6)
True
```

**Comment by toshas** [comment]
@IvanYashchuk @mruberry May I suggest the following example be added to the function documentation, given that parameterization functionality is merged now? It demonstrates the simplest usage of `torch.linalg.householder_product` to implement exact orthogonality constraints in linear layers and directs the reader to Sec.7 of our recent publication: https://arxiv.org/abs/2103.04217 (which was the basis of this feature request). It elaborates how to make use of the batch dimension, parameterize multiple matrices of different shapes within the same batch, and other use cases of `torch.linalg.householder_product`.

```
The second example shows the orthonormal parameterization of a weight matrix.
Parameterizing multiple matrices of different sizes in a single batch and reduced parameterizations 
are discussed in Sec.7 of https://arxiv.org/abs/2103.04217.

Example 2: Exact orthonormal constraints on weight matrices:

>>> class Orthonormal(torch.nn.Module):
>>>    def __init__(self, d, r):
>>>        super().__init__()
>>>        self.register_buffer('eye', torch.eye(d, r))
>>>    def forward(self, hh):
>>>        hh = hh.tril(-1) + self.eye
>>>        tau = 2 / (hh * hh).sum(dim=0)
>>>        return torch..linalg.householder_product(hh, tau)
>>>
>>> cin, cout = 128, 256
>>> layer = torch.nn.Linear(cin, cout)
>>> torch.nn.utils.parametrize.register_parametrization(layer, 'weight', Orthonormal(cout, cin))
ParametrizedLinear(
  in_features=128, out_features=256, bias=True
  (parametrizations): ModuleDict(
    (weight): ParametrizationList(
      (0): Orthonormal()
    )
  )
)
>>> torch.allclose(layer.weight.T.mm(layer.weight), torch.eye(cin), atol=1e-6)
True
```

**Comment by mruberry** [comment]
Hey @toshas, that's an interesting idea. cc @jbschlosser @albanD

Not every pattern is a great candidate for our documentation, in part because it can be difficult to discover when put on the pytorch.org docs website. This particular pattern may be better in a blog you write, for example, that would give you the opportunity to talk more about your research, too.
